[Doc Feature][Copilot Handoff] (#10851)
MicrosoftDocs/msteams-docs: msteams-platform/bots/how-to/conversations/bot-copilot-handoff.md

---
title: Copilot handoff
description: Learn how to allow users to continue their conversation with custom engine copilot from the copilot for Microsoft 365.
ms.date: 05/07/2024
ms.topic: conceptual
author: v-ypalikila
ms.author: surbhigupta
---

# Copilot handoff

> You can enable copilot handoffs for your plugins to allow users to continue the conversation with a custom engine copilot from Microsoft Copilot for Microsoft 365 without losing the context of their work. Copilot handoffs allow for a seamless transition from Copilot for Microsoft 365 to your custom engine copilot by using deep links that carry over the user context.

Reviewer: ```suggestion
You can enhance your Copilot message extension plugin to hand off a conversation to your [custom engine copilot](/microsoft-365-copilot/extensibility/overview-custom-engine-copilot) to handle cases where a more in-depth, specialized chat experience is required. Copilot handoffs are plugin-provided deep links that carry over chat context, enabling users to seamlessly transition their chat with Copilot for Microsoft 365 to your bot service.
```

> In the following graphic, a user looks for help on tech issues using copilot and receives results from the enterprise dataset. The user finds the information is insufficient and wants to continue the interaction with a Contoso bot plugin.​ When the user selects the Contoso bot, a new chat starts and the conversation continues in the Contoso bot. This seamless switch from copilot to the Contoso bot, without losing the conversation context is called copilot handoff.

Reviewer: ```suggestion
In the following graphic, a user looks for help on tech issues using Copilot and receives results from the enterprise dataset. The user finds the information is insufficient and wants to continue the interaction with a Contoso bot plugin.​ When the user selects the Contoso bot, a new chat starts and the conversation continues in the Contoso bot. This seamless switch from Copilot to the Contoso bot, without losing the conversation context is called a *copilot handoff*.
```

:::image type="content" source="../../../assets/images/Copilot/Copilot-handoff.gif" alt-text="The GIF shows the conversation handoff between the copilot for Microsoft 365 and the Contoso chat bot.":::

## How copilot handoff works

> A custom engine copilot sends a deep link with a continuation token to the copilot for Microsoft 365. The deep link query parameter with a continuation token ensures that any information from your plugin invocation parameters is referenced. When the user selects the deep link, the copilot then sends an invoke call to the bot with the continuation token, and the bot can resume the conversation based on the context. This process enables a seamless transition from copilot to a custom engine copilot, maintaining the conversation’s continuity and context, which optimizes the user experience.

Reviewer: ```suggestion
A custom engine copilot sends a deep link with a continuation token to Copilot for Microsoft 365. The deep link query parameter with a continuation token ensures that any information from your plugin invocation parameters is referenced. When the user selects the deep link, Copilot then sends an invoke call to the bot with the continuation token, and the bot can resume the conversation based on the context. This process enables a seamless transition from Copilot for Microsoft 365 to your custom engine copilot, maintaining the conversation’s continuity and context, which optimizes the user experience.
```

:::image type="content" source="../../../assets/images/Copilot/copilot-handoff-flow.png" alt-text="Screenshot shows the handoff flow between the user, Copilot, plugin, Teams, and bot." lightbox="../../../assets/images/Copilot/copilot-handoff-flow.png":::

You must create a deep link URL with a `continuation` query parameter for the action button and assign a continuation token to the parameter to facilitate the handoff process. When the user selects the action button, Microsoft Teams reads the continuation token from the URL and initiates an invoke call to the bot. The bot or plugin uses the continuation token to create a response, which is then displayed to the user in the plugin chat window.

## Enable copilot handoff

To enable copilot handoff in Teams, follow these steps:

> 1. **Configure a deep link URL**: Create a [deep link to a chat](../../../concepts/build-and-test/deep-link-teams.md#configure-deep-link-to-start-a-chat-manually) and add the `28:<botId>` and `continuationToken` to the deep link format. The deep link format must be `https://teams.microsoft.com/l/chat/0/0?users=${28:<botId>}&continuation=${continuationToken}`.

Reviewer 2: bring `28:` outside the ${} in https://teams.microsoft.com/l/chat/0/0?users=${28:<botId>}&continuation=${continuationToken}  i.e https://teams.microsoft.com/l/chat/0/0?users=28:${botId}&continuation=${continuationToken}
Same comment for Example below at line 46

Author: Updated.

---
title: Copilot handoff
description: Learn how to allow users to continue their conversation with custom copilots from the Copilot for Microsoft chat.
ms.date: 05/07/2024
ms.topic: conceptual
author: v-ypalikila
ms.author: surbhigupta
---

# Copilot handoff

> You can enable Copilot handoff to allow users to continue the conversation with a custom Copilot from the Copilot for Microsoft chat without losing the context of their work. You can improve your plugins with the Copilot handoff feature, which allows for a seamless transition to your custom copilot by using deep links that carry over the user context. This resolves the issue of users having to restart and repeat their requests in the bot experience. For example, in the following image, the user is using Copilot to get information on tech issues and receives results from across the enterprise dataset. The user doesn't find the information sufficient and wants to continue an interaction with an Aisera bot.​ When the user selects the Aisera bot, a new chat is opened to continue the conversation with Aisera. The continuation of the conversation without losing the context is called Copilot handoff.

Reviewer 3: Please break up and simplify this wall of text, as it is hard to read.

Author: Updated the Introduction and split the intro and examples into two paragraphs.

:::image type="content" source="../../../assets/images/Copilot/copilot-handoff-flow.png" alt-text="Screenshot shows the handoff flow between the user, Copilot, plugin, Teams, and bot." lightbox="../../../assets/images/Copilot/copilot-handoff-flow.png":::

You must create a deep link URL with a `continuation` query parameter for the action button and assign a continuation token to the parameter to facilitate the handoff process. When the user selects the action button, Microsoft Teams reads the continuation token from the URL and initiates an invoke call to the bot. The bot or plugin uses the continuation token to create a response, which is then displayed to the user in the plugin chat window.

## Enable copilot handoff

To enable Copilot handoff in Teams, follow these steps:

1. **Configure a deep link URL**: Create a deep link to the bot chat and include `botMri` and `continuationToken`. The deep link must be in the format `https://teams.microsoft.com/l/chat/0/0?users=${botMri}&continuation=${continuationToken}`.

   **Example**:

   ```JSON
   { 

   "type": "Action.OpenUrl", 

   "title": "Handoff to Bot", 

>    "url": "https://teams.microsoft.com/l/chat/0/0?users=${botMri}&continuation=${continuationToken}" 

Reviewer 3: > https://teams.microsoft.com/l/chat/0/0?users=${botMri}

This deeplink format isn't documented, please document (the docs do not say that you can use botMri for users).

If I remember correctly it was previously documented, but then we removed it because we didn't want using this deep link to start a chat with the bot. Have we changed our mind, and this is ok again?

Author: Updated as Create a [deep link to a chat](https://learn.microsoft.com/en-us/microsoftteams/platform/concepts/build-and-test/deep-link-teams#configure-deep-link-to-start-a-chat-manually) and add the `28:<botId>` and `continuationToken` to the deep link format."

   }
   ```

   **Parameters**

   |Property |Description  |
   |---------|---------|
   |`botMri`     |The `botMri` must be in the format `28:<botID>`. The botMri specifies the bot chat that the user is redirected to when they select the action button that contains the deep link url.        |
   |`continuationToken`     | The `continuationToken` value can be any string, but ensure that the total URL length doesn't exceed 2048 characters and the string is URL encoded.        |
   |`Action.OpenUrl`     |  The `Action.OpenUrl` property allows the user to handoff the conversation to a bot.       |

>    **Sample payload**

Reviewer 3: What is this a sample payload of? It seems to be of the `handoff/action` invoke, but you haven't introduced that invoke type yet.

Author: Update as "The bot receives the continuation token from the URL in the form of an invoke and uses the token to maintain the context of the conversation."

        "tenantId": "72f988bf-86f1-41af-91ab-2d7cd011db47", 
        "id": "a:13tOiSzRqeub3zaqoTHKpvOkk8Y1zFxk-g8WKdAUM2tjhTBFMt4RSuL8YWi7uwFNBmbxsyzYYktJEyfimYXYiEoplQ34aJs1y8trDb7EIcG09xOjSUieHVzFZ2b8tkagZ" 
    }, 
    "recipient": { 
        "id": "28:68935e91-ff09-4a33-a675-0fe09f015706", 
        "name": "NorthwindProducts" 
    }, 
    "entities": [ 
        { 
            "locale": "en-US", 
            "country": "US", 
            "platform": "Android", 
            "timezone": "America/Chicago", 
            "type": "clientInfo" 
        } 
    ], 
    "channelData": { 
        "tenant": { 
            "id": "72f988bf-86f1-41af-91ab-2d7cd011db47" 
        }, 
        "source": { 
            "name": "message" 
        }, 
        "legacy": { 
            "replyToId": "1:1_qLAAGcfze29QAWxzicc7gvR3vuNAlKvth08vavxYYs" 
        } 
    }, 
    "replyToId": "1713210583687", 
    "value": { 
        "continuation": "test-continuation-token" 
    }, 
    "locale": "en-US", 
    "localTimezone": "America/Chicago", 
    "rawTimestamp": "2024-04-15T19:50:32.945Z", 
    "rawLocalTimestamp": "2024-04-15T14:50:32.945-05:00", 
    "callerId": "urn:botframework:azure" 
   }
   ```

>    The `Action.OpenUrl` property allows the user to handoff the conversation to a bot.  After the user is redirected to the bot chat, the bot receives the `continuationToken` value as an invoke message of the type `handoff/action`, if the app associated with the bot is installed in the user's personal scope and the continuationToken value isn't empty.

Reviewer 3: "hand off", since it's used as a verb here.

Author: updated.

>    The `Action.OpenUrl` property allows the user to handoff the conversation to a bot.  After the user is redirected to the bot chat, the bot receives the `continuationToken` value as an invoke message of the type `handoff/action`, if the app associated with the bot is installed in the user's personal scope and the continuationToken value isn't empty.

Reviewer 3: In what context is this invoke sent? Let's not make the developer pore over the sample payload to answer key questions.

Author: Updated as "When a user selects on the action button, the deep link is activated and opens a new chat window with the bot."  (which is 1:1 chat with bot).

>    The `Action.OpenUrl` property allows the user to handoff the conversation to a bot.  After the user is redirected to the bot chat, the bot receives the `continuationToken` value as an invoke message of the type `handoff/action`, if the app associated with the bot is installed in the user's personal scope and the continuationToken value isn't empty.

Reviewer 3: What happens if the app isn't installed?

> 1. **Handle the invoke type**: In your bot code, manage the `handoff/action` invoke using the `onInvokeActivity` handler. Override the `onInvokeActivity` method to handle the invoke call. The response to this invoke call must be an HTTP status code 200 for success or a code in the range of 400-500 for error.

Reviewer 3: What happens for other status codes?

Are we strictly saying 400-500 only? What about 502? 503?

Reviewer 3: What does the user see if an error status code is returned?

Reviewer: How about `or a (400- or 500-level) error code response` (in place of "or a code in the range of 400-500 for error")

   > [!NOTE]
   > Don't send any payload with this response as it doesn't render in the chat window. The responses based on the continuation token must be sent to the user separately.

>    Here's an example of how to handle the invoke call in `searchApp.ts`:

Reviewer 3: Not sure what this example shows. L120-122 are the only relevant parts, and the heart is a TODO.

Author: Removed the code which is not revelant.

          };
        }
      }
   ```

   When the bot receives the invoke call, `context.activity.value.continuation` contains the `continuationToken` that was set in the deep link URL.

## Best practices

* It’s important for the bot to quickly notify users when they're directed to the bot chat after selecting the action button, as there's no visual indication. This helps manage expectations as there might be a delay before the bot returns a response due to network latency and processing time. For instance, the bot can send a series of activities to keep the user informed of the progress:

    ```typescript
    await context.sendActivities([
      {
        type: ActivityTypes.Message,
        text: "Continuing conversation from Copilot...",
      },
      { type: ActivityTypes.Typing },
      { type: "delay", value: 1000 },
      {
        type: ActivityTypes.Message,
        text: `Fetching more details using the continuation token passed: ${continuationToken}`,
      },
      { type: ActivityTypes.Typing },
      { type: "delay", value: 4000 },
      {
        type: ActivityTypes.Message,
        text: `Handoff successful!`,
      }
    ]);
    
    ```

> * For a smooth handoff process, it’s important to manage the continuation token effectively. Keep the token’s lifetime short and ensure it doesn't process more than once. After a token is used, remove it from storage to prevent it from being used again. If the same token comes up, let the user know they need to start a new conversation with the bot because the handoff from Copilot can't continue.

Reviewer 3: Can we explain why it's "important" to have a short lifetime, and ensure that the token is only redeemed once? If the action is idempotent why is this necessary?

Author: As discussed with Gino and prem, updated a generic guidance to state that developers must handle the token expiry and multiple reuses of the token.

"We recommended that you manage the lifecycle of the continuation token to ensure it expires after a reasonable period and also handle scenarios where the continuation token request is replayed by the user."

> * For a smooth handoff process, it’s important to manage the continuation token effectively. Keep the token’s lifetime short and ensure it doesn't process more than once. After a token is used, remove it from storage to prevent it from being used again. If the same token comes up, let the user know they need to start a new conversation with the bot because the handoff from Copilot can't continue.

Reviewer 3: I think we should say something about the continuation token. That it's the developer's responsibility to implement. What an implementation might look like. For an app that is implementing handoff, this I suspect that's going to be the crux. The rest of this is basically a simple "bookmark" mechanism that lets you pass state from one conversation to another.

Author: Updated a general guidance that we recommend developers to handle the token expiry and reusing the same token.

> * When the bot is unable to process a request or if the token is expired, it must actively guide the user to the bot chat. If the conversation can't continue, the bot must inform the user of the error and suggest they start a new conversation or retry their query.

Reviewer 3: Is the recommendation in this case different from the one in the previous bullet point? I'm reading and re-reading the two and I'm not sure I understand the difference.

Author: removed the second bullet point.
